src/atlas/interpolation/element/Quad2D.cc

```diff
@@ -51,37 +51,55 @@ method::Intersect Quad2D::intersects(const PointXY& r, double edgeEpsilon, doubl
 method::Intersect Quad2D::localRemap(const PointXY& p, double edgeEpsilon, double epsilon) const {
     method::Intersect isect;
 
+    // get area of quad.
+    double quadArea = area();
+
+    // Epsilon is compared against areas. Scale value accordingly.
+    double areaEpsilon = epsilon * quadArea;
+
     // work out if point is within the polygon
-    if (!inQuadrilateral({p.x(), p.y()})) {
+    if (!inQuadrilateral({p.x(), p.y()}, areaEpsilon)) {
         return isect.fail();
     }
 
-    auto solve_weight = [epsilon](const double a, const double b, const double c, double& wght) -> bool {
-        if (std::abs(a) > epsilon) {
-            // if a is non-zero, we need to solve a quadratic equation for the weight
-            // ax**2 + bx + x = 0
+    auto solve_weight = [&](double a, double b, double c, double& wght) -> bool {
+
+        auto check_weight = [](double wght, double tol) -> bool {
+            return ((wght > -tol) && (wght < (1. + tol)));
+        };
+
+        if (std::abs(a) > areaEpsilon) {
+            // if a is non-zero, we need to solve a quadratic equation for the weight.
+            // ax**2 + bx + c = 0
             double det = b * b - 4. * a * c;
-            if (det >= 0.) {
+
+            // det has dimensions of area squared.
+            if (det > -areaEpsilon * 2. * quadArea) {
+
                 double inv_two_a = 1. / (2. * a);
-                double sqrt_det  = std::sqrt(det);
-                double root_a    = (-b + sqrt_det) * inv_two_a;
-                if ((root_a > -epsilon) && (root_a < (1. + epsilon))) {
-                    wght = root_a;
-                    return true;
-                }
-                double root_b = (-b - sqrt_det) * inv_two_a;
-                if ((root_b > -epsilon) && (root_b < (1. + epsilon))) {
-                    wght = root_b;
-                    return true;
-                }
+                double sqrt_det  = std::sqrt(std::max(det, 0.));
+
+                // x = A +/- B,
+                // A = -b / 2a,
+                // B sqrt_det / 2a.
+
+                double A = -b * inv_two_a;
+                double B = sqrt_det * inv_two_a;
+
+                // Scale epsilon by magnitude of A or B.
+                double edgeEpsilonScaled = edgeEpsilon *
+                                           std::max({1., std::abs(A), std::abs(B)});
+
+                return check_weight(wght = A + B, edgeEpsilonScaled) ? true :
+                       check_weight(wght = A - B, edgeEpsilonScaled) ? true : false;
```

Comment:
It looks really elegant to have "wght = A + B" or "wght = A - B" inline, but it is a bit confusing. I'm sure you were thinking it too 😉.
Also, perhaps we could spell "wght" out as "weight" 😏